File: Algorithm/ExpandAroundCenter/leetcode/5. Longest Palindromic Substring/GPT/Longest_Palindromic_Substring.py
# ### 1. 問題分析結果

# #### 競技プログラミング視点

# * **目的**: 文字列 `s` の最長回文部分文字列を返す。
# * **制約**: `1 <= |s| <= 1000`, 文字種は英字・数字。
# * **最速手法の選択**: 入力サイズが小さめなため、**中心展開法（Expand Around Center）**で十分。最悪 O(n²)、追加メモリ O(1)、実装が短く分岐も少ない。
# * **メモリ最小化**: 展開中に部分文字列を生成せず、**インデックスのみ保持**。最後に一度だけ `s[start:end+1]` を返す。

# #### 業務開発視点

# * **型安全性/可読性**: `pylance` での静的解析に耐える **厳密な型ヒント**。ヘルパ関数の引数・戻り値に `tuple[int, int]` を付与。
# * **エラーハンドリング**: 実運用を想定し、制約違反を検出する **安全版メソッド** を別途用意（LeetCode 本戦用メソッドは高速性優先でバリデーション省略）。
# * **保守性**: 展開処理を小関数に分離、責務を明確化し再利用可能に。

# #### Python特有分析

# * **CPython最適化**:

#   * `while` ループと添字比較（`s[l] == s[r]`）で**C実装の文字比較**を活用。
#   * ループ内ではオブジェクト生成を避け、**インプレースにインデックス更新**。
# * **データ構造**: 追加構造は不要（O(1)）。`list`/`deque` は使用しない。
# * **GIL**: 単一スレッドの CPU 計算であり、GILは実質影響なし。

# ---

# ### 2. アルゴリズム比較表

# | アプローチ         | 時間計算量 | 空間計算量 | Python実装コスト | 可読性 | 標準ライブラリ活用 | CPython最適化   | 備考                |
# | ------------- | ----- | ----- | ----------- | --- | --------- | ------------ | ----------------- |
# | 方法A: Manacher | O(n)  | O(n)  | 高           | 中   | 不要        | 適（連続配列走査）    | 最速だが境界処理が複雑       |
# | 方法B: DP       | O(n²) | O(n²) | 中           | 中   | 不要        | 普通           | テーブル埋めで直感的だがメモリ重い |
# | 方法C: 中心展開（採用） | O(n²) | O(1)  | 低           | 高   | 不要        | 適（単純な while） | n≤1000 なら十分高速     |

# ---

# ### 3. 採用アルゴリズムと根拠

# * **選択**: 方法C（中心展開）
# * **根拠**:

#   * 入力上限 1000 に対して O(n²) は実用十分、**追加メモリ O(1)**。
#   * コードが短く、**バグ混入率が低い**。レビュー容易。
#   * Python の文字比較は C 実装で高速。**スライス生成は最後の 1 回のみ**。
# * **Python最適化戦略**:

#   * ループは `for i in range(n)` と `while` を使用（関数呼び出し最小化）。
#   * ループ内で**一時サブストリングを作らない**。
#   * ローカル変数束縛（`best_l`, `best_r`）で **LOAD_FAST** を活用。

# ---

# ### 3. 実装パターン

# * **競技版**: `Solution.longestPalindrome`（LeetCodeで採点される想定・入力検証を省略し最速重視）。
# * **安全版**: `Solution.longestPalindrome_safe`（制約検証と例外送出を含む）。本戦では使われないが、業務での再利用を想定。

# ---

# ### 4. 検証（方針のみ）

# * **境界**: 長さ1（例: `"a"`）、全同一（`"aaaa"`）、偶数解（`"cbbd"→"bb"`）、複数解（`"babad"→"bab"/"aba"`）、先頭/末尾が解となるケース。
# * **型チェック**: `pylance` 下で型不一致が出ない注釈（`tuple[int, int]`、戻り値 `str`）。

# ---

# ### 実装コード（LeetCode クラス形式・型注釈付き）

# ```python
# from __future__ import annotations

# from typing import Final, Tuple


# class Solution:
#     """
#     Longest Palindromic Substring

#     競技プログラミング向け（最速・最小メモリ）と、
#     業務向けのバリデーション込みメソッドを提供。
#     """

#     def longestPalindrome(self, s: str) -> str:
#         """
#         LeetCode 採点対象想定：高速版
#         Time: O(n^2), Space: O(1)

#         競技用途のため入力検証は省略（問題制約を前提）。
#         """
#         n: int = len(s)
#         if n <= 1:
#             return s

#         # ローカル関数にしてクロージャアクセスを高速化（LOAD_FAST）
#         def expand(l: int, r: int) -> Tuple[int, int]:
#             # [l, r] から可能な限り拡張し、inclusive 範囲を返す
#             while l >= 0 and r < n and s[l] == s[r]:
#                 l -= 1
#                 r += 1
#             return l + 1, r - 1

#         best_l: int = 0
#         best_r: int = 0

#         # 各位置を中心に奇数長・偶数長の両方を展開
#         for i in range(n):
#             l1, r1 = expand(i, i)       # 奇数長
#             if r1 - l1 > best_r - best_l:
#                 best_l, best_r = l1, r1

#             l2, r2 = expand(i, i + 1)   # 偶数長
#             if r2 - l2 > best_r - best_l:
#                 best_l, best_r = l2, r2

#         # 最後に一度だけスライスを生成（GC/メモリ負荷を最小化）
#         return s[best_l : best_r + 1]

#     # ---- 以下は業務開発向けの安全版（必要に応じて使用） ----
#     def longestPalindrome_safe(self, s: str) -> str:
#         """
#         入力検証付きの安全版。
#         問題制約に違反する入力を受け取る可能性がある場合はこちらを使用。

#         Raises:
#             TypeError: s が文字列でない場合
#             ValueError: 長さが [1, 1000] を外れる場合
#             ValueError: 英数字以外を含む場合
#         """
#         self._validate_input(s)
#         return self.longestPalindrome(s)

#     @staticmethod
#     def _validate_input(s: object) -> None:
#         if not isinstance(s, str):
#             raise TypeError("Input must be a string")
#         n = len(s)
#         if n < 1 or n > 1000:
#             raise ValueError("Input length must be within [1, 1000]")
#         # 英字・数字のみ（問題制約）
#         # Python の re を使わず、全体1パスでチェック（余計な依存なし）
#         for ch in s:
#             # isalnum は Unicode 対応だが、英字/数字以外を弾くため ASCII 範囲を厳密判定
#             o = ord(ch)
#             is_digit = 48 <= o <= 57       # '0'-'9'
#             is_upper = 65 <= o <= 90       # 'A'-'Z'
#             is_lower = 97 <= o <= 122      # 'a'-'z'
#             if not (is_digit or is_upper or is_lower):
#                 raise ValueError("Input must consist of digits and English letters only")
# ```

# **補足（pylance 対応）**

# * すべての公開メソッドに戻り値の型ヒント `-> str` を付与。
# * ヘルパ `expand` は `Tuple[int, int]` を明示し、インデックスは `int` に限定。
# * 依存は標準のみ・外部ライブラリ不使用。

# * `from __future__ import annotations` が**ファイル最上行でない**（LeetCode 側の前置コードや不可視文字の混入で先頭扱いでなくなる）
# * Markdown からのコピーで **不可視文字（ゼロ幅スペース等）** が `def` 行や直前行に混入

# どちらも本質ロジックとは無関係です。確実に通るように、**将来インポートや型用インポートを一切使わない純粋なクラス実装**にリライトしました（ASCII のみ・4スペースインデント）。

# ---

# ## 修正後コード（LeetCode / Class 形式, 型注釈あり, 追加依存なし）

# ```python
class Solution:
    def longestPalindrome(self, s: str) -> str:
        """
        Longest Palindromic Substring
        Time: O(n^2), Space: O(1)
        """
        n: int = len(s)
        if n <= 1:
            return s

        def expand(l: int, r: int) -> tuple[int, int]:
            # 中心 (l, r) から最大まで拡張し、inclusive の [L, R] を返す
            while l >= 0 and r < n and s[l] == s[r]:
                l -= 1
                r += 1
            return l + 1, r - 1

        best_l: int = 0
        best_r: int = 0

        for i in range(n):
            l1, r1 = expand(i, i)  # 奇数長
            if r1 - l1 > best_r - best_l:
                best_l, best_r = l1, r1

            l2, r2 = expand(i, i + 1)  # 偶数長
            if r2 - l2 > best_r - best_l:
                best_l, best_r = l2, r2

        return s[best_l : best_r + 1]


# ```

# ### 変更点（エラー回避のための対策）

# * `from __future__ import annotations` と `typing` インポートを**完全削除**
#   → 3.11 では不要。先頭に置けない環境や不可視文字混入の影響を排除。
# * すべて **ASCII** のみ／**4スペースインデント**で整形
#   → ゼロ幅スペースやタブ混在による構文エラーを回避。
# * 型注釈は **ビルトインの `tuple[int, int]`** を使用（追加 import 不要）。

# このまま LeetCode のエディタに貼り付ければ `SyntaxError` は解消され、正答が返るはずです。

File: Algorithm/ExpandAroundCenter/leetcode/5. Longest Palindromic Substring/GPT/Longest_Palindromic_Substring.js
// # 1. 問題の分析

// ## 競技プログラミング視点での分析

// * 文字列 `s` の最長回文部分文字列を返す。`|s| ≤ 1000` なので **O(n²), O(1)** の中心展開法で十分速い。
// * Manacher（O(n)）も可だが実装コストが高い割にこの制約では恩恵が小さい。

// ## 業務開発視点での分析

// * 読みやすさと保守性重視。**中心展開を小関数に分離**して責務を明確化。
// * 例外は**関数入口で早期に検証**し、ホットパスから排除。

// ## JavaScript特有の考慮点

// * **V8 最適化**：インデックス型 `for`、プリミティブと固定プロパティの利用。
// * **GC 対策**：展開中に `substring` を生成しない（最終結果だけ作成）。
// * **クロージャ最小化**：`expand` はスコープ安定、引数はプリミティブのみ。

// ---

// # 2. アルゴリズムアプローチ比較

// | アプローチ        | 時間計算量 | 空間計算量 | JS実装コスト | 可読性 | 備考                |
// | ------------ | ----- | ----- | ------- | --- | ----------------- |
// | 方法A：Manacher | O(n)  | O(n)  | 高       | 低   | 最速だがバグりやすい境界処理    |
// | 方法B：DP       | O(n²) | O(n²) | 中       | 中   | 表を埋めるがメモリ重い       |
// | 方法C：中心展開（採用） | O(n²) | O(1)  | 低       | 高   | 実装が短く、n≤1000で十分速い |

// ※「JS実装コスト」＝型安定性・境界処理の難度・コード量を総合評価。

// ---

// # 3. 選択したアルゴリズムと理由

// * **選択**：方法C（中心展開）
// * **理由**：O(n²) でも n≤1000 で実用十分。コード短小でレビュー容易。GC 圧を抑えられる。
// * **JS最適化ポイント**：

//   * ループは `for (let i=0; i<n; i++)` 固定。
//   * 走査中は**開始・終了インデックスのみ更新**し、最後に一度だけ `substring`。
//   * 例外チェックは冒頭で完結。

// ---

// # 4. コード実装（solution.js）

// ```javascript
'use strict';

/**
 * Longest Palindromic Substring
 * LeetCode 形式：純関数・I/Oなし
 *
 * 実行環境:
 * - Language/Runtime: JavaScript (Node.js v22.14.0) ※互換: v18+
 * - Module: CommonJS
 * - 外部ライブラリ: 不使用（Node 標準のみ）
 *
 * @function longestPalindrome
 * @param {string} s - 入力文字列（英数字のみ）
 * @returns {string} 最長回文部分文字列
 * @throws {TypeError} s が文字列でない
 * @throws {RangeError} 長さが [1, 1000] を外れる
 * @throws {RangeError} 非許可文字（英字・数字以外）が含まれる
 * @complexity 時間 O(n^2), 空間 O(1)
 */
function longestPalindrome(s) {
    // ---- 入力検証（早期・軽量）----
    if (typeof s !== 'string') throw new TypeError('Input must be a string');
    const n = s.length;
    if (n < 1 || n > 1000) throw new RangeError('Input length must be in [1, 1000]');
    // 問題文の制約に合わせ英数字のみを許可（余計なユニコード正規化等は行わない）
    // 正規表現は一度だけ生成し、隠れクラスを安定化
    const re = /^[A-Za-z0-9]+$/;
    if (!re.test(s)) throw new RangeError('Input must consist of digits and English letters only');

    // ---- 中心展開のヘルパ（クロージャ軽量）----
    /**
     * 中心 (l,r) から最大まで拡張し、拡張後の [L, R] を返す。
     * @param {number} l
     * @param {number} r
     * @returns {[number, number]} inclusive 範囲
     */
    function expand(l, r) {
        // while で境界・一致を確認。ループ内は読み取りのみで型安定。
        while (l >= 0 && r < n && s.charCodeAt(l) === s.charCodeAt(r)) {
            l--;
            r++;
        }
        // 1つ進み過ぎているので戻す（inclusive）
        return [l + 1, r - 1];
    }

    // ---- 本処理：各インデックスを中心として奇数長・偶数長を展開 ----
    let bestL = 0,
        bestR = 0; // inclusive
    for (let i = 0; i < n; i++) {
        // 奇数長（単一中心）
        const odd = expand(i, i);
        if (odd[1] - odd[0] > bestR - bestL) {
            bestL = odd[0];
            bestR = odd[1];
        }
        // 偶数長（隣接2文字中心）
        const even = expand(i, i + 1);
        if (even[1] - even[0] > bestR - bestL) {
            bestL = even[0];
            bestR = even[1];
        }
    }

    // 返却時のみ substring を生成（GC 圧低減）
    return s.substring(bestL, bestR + 1);
}

module.exports = { longestPalindrome };
// ```

// ---

// # 5. 追加メモ（JS最適化チェックリスト）

// * ループは `for` 基本。`forEach/map` は割当コスト増＆型崩し要因になりうるため不使用。
// * 一時オブジェクト生成は最小（結果のみ `substring`）。展開は数値インデックスで管理。
// * `charCodeAt` 比較により高速な等価判定（`s[i]` でも可だが、古い環境互換と微最適化）。
// * 例外はエントリで早期に投げ、ホットパスから分離。
// * Pure function（グローバル書き換えなし・入力不変）。

// ---

// ## 0. 実行環境（要件トレース）

// * Language/Runtime: **JavaScript (Node.js v22.14.0 / 互換 v18+)**
// * Module: **CommonJS**
// * 外部ライブラリ: **禁止（標準のみ）**
// * CI 前提: `node solution.js` は**副作用なし**のため、テスト側から `require('./solution').longestPalindrome` を呼び出す運用を想定。
